### PR TITLE
rest api design document created

### PR DESCRIPTION
**Pull request by perrymant**
1. `Layout`: I had a hard time making "readable/aesthetically pleasing" markdown output, because of conflicts with how the JSON is laid out. It's almost better reading the raw markdown as the document. I wonder if you know how I could better lay this out.

2. `Content`: Besides layout, I hope I have included the correct information for the API design. Can you point me to anything I'm missing. The sprint task is below for convenience.

3. `Learning Notes`: I've also made some learning notes on API design (This resides in my Learning Notes repo at: https://github.com/perrymant/learning-notes/blob/master/17-rest-api-design.md.) Should I have branched this new document and PR'ed it rather than just add it to master?

Sprint task: "API design. Identify the functionalities of my current CLI and map these to HTTP REST endpoints. What should the URL look like, the HTTP methods, what is the structure of the JSON for both requests and responses. Look at server codes, and put these in the design."

### CODE REVIEW

File: design/api-specification-design.md [hunk 1]
```diff
@@ -0,0 +1,157 @@
+# api-specification-design
+Personal notes about REST API design [here](https://github.com/perrymant/learning-notes/blob/master/17-rest-api-design.md)
+
+## HTTP Headers for all methods:
+Content-Type: application/json
+Accept: application/json
+
+## 1. help
```
Comment (kinbiko) [inline design/api-specification-design.md]: Help is not something you care about for HTTP APIs really - only CLIs: CLIs may are often used directly by the user, whereas a HTTP API is used by another program and there is rarely anything useful to gain from a program telling its user the help instructions from another program (the API).

Let's simplify to only have report and adding a new transaction

File: design/api-specification-design.md [hunk 2]
```diff
@@ -0,0 +1,157 @@
+# api-specification-design
+Personal notes about REST API design [here](https://github.com/perrymant/learning-notes/blob/master/17-rest-api-design.md)
+
+## HTTP Headers for all methods:
+Content-Type: application/json
```
Comment (kinbiko) [inline design/api-specification-design.md]: I wonder if this is an unnecessary abstraction - keep all the headers expected for the requests and responses along with their respective bodies.

Case in point: GET requests don't have the content-type header set, typically

File: design/api-specification-design.md [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+        ║ 2018-01-02 │ DEBIT            │ £0.72  │ £1.78   │ Paid bill   ║
+        ╟────────────┼──────────────────┼────────┼─────────┼─────────────╢
+        ║ 2018-01-01 │ CREDIT           │ £1.25  │ £2.50   │ Got paid    ║
+        ╟────────────┼──────────────────┼────────┼─────────┼─────────────╢
+        ║ 2018-01-01 │ CREDIT           │ £1.25  │ £1.25   │ Got paid    ║
+        ╚════════════╧══════════════════╧════════╧═════════╧═════════════╝
+```
+
+## 2. report
+The user asks for `report` and obtains the report of their account.
```
Comment (kinbiko) [inline design/api-specification-design.md]: We still don't have a notion of a user, do we? Let's explicitly assume that all users are using the same account then for the purposes of this document.

File: design/api-specification-design.md [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+
+#### Request
+| Method | URL |
+|--------|-----|
+| GET   | api/report/ |
+
+#### Response
+- 200: a `report statement` will be sent:
+
+`{
```
Comment (kinbiko) [inline design/api-specification-design.md]: Use triple backticks with a filetype of JSON for better formatting.

Remember that JSON != JavaScript objects, and all your keys (LHS of `:`) will have to be `"strings"`.

Applies throughout.

File: design/api-specification-design.md [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+    status : "success",
+    data : {
+        "report" : "---report statement---"
```
Comment (kinbiko) [inline design/api-specification-design.md]: what does the report statement look like?

File: design/api-specification-design.md [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+     }
+}`
+
+- 400:  ">Error: Invalid argument: `moneymaker help` >for more info.":
```
Comment (kinbiko) [inline design/api-specification-design.md]: How can we have an invalid argument when hitting `/api/report`? There are no arguments.

I think this 400 is more applicable for `transaction`.

File: design/api-specification-design.md [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+`{
+    "status" : "fail",
+    "data" : { "Error" : "Error: Invalid argument:\nmoneymaker help \nfor more info." }
+}`
+
+- 500:  "Something went wrong. Please try again later.":
+
+`{
+    "status" : "error",
+    "message" : "Unable to communicate with database"
```
Comment (kinbiko) [inline design/api-specification-design.md]: The client of a REST API should *never* be informed of the internals of what went wrong. Can you think of why?

In this case we should just return the same generic error message each time.
> "Something went wrong. Please try again later."
is perfect

File: design/api-specification-design.md [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+#### Request
+| Method | URL |
+|--------|-----|
+| GET   | api/report/ |
+
+#### Response
+- 200: a `report statement` will be sent:
+
+`{
+    status : "success",
```
Comment (kinbiko) [inline design/api-specification-design.md]: I'm not sure this `status` key offers any more insight than the status code does. Consider getting rid of it.

File: design/api-specification-design.md [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+| Method | URL |
+|--------|-----|
+| PUT   | api/transaction-amount-description-date/ |
+
+#### Response
+
+- 200: A `transaction statement` will be sent:
+`{
+    status : "success",
+    data : {
```
Comment (kinbiko) [inline design/api-specification-design.md]: What's the reasoning behind having this `data` wrapper key? Why not just return

```json
{
  "transactions": [...]
}
```
directly?

File: design/api-specification-design.md [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+
+`report statement` for status code 200:
+
+## 3. transaction
+The user makes a transaction and sends it to update their account.
+
+#### Request
+| Method | URL |
+|--------|-----|
+| PUT   | api/transaction-amount-description-date/ |
```
Comment (kinbiko) [inline design/api-specification-design.md]: I think POST is the best verb to use here. POST used when creating something new. PUT is generally used for updates.

Comment (kinbiko) [inline design/api-specification-design.md]: This is an awkward URL. I'd expect this to just be `POST api/transaction` really.
If we have to list all the arguments we expect then this will probably not age well, in that adding more transactions means changing the URL. Can you think of a reason why this might not be desirable?

File: design/api-specification-design.md [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+             ...
+           ]
+           }
+}`
+
+- 400:  ">Error: Invalid argument: `moneymaker help` >for more info.":
+
+`{
+    "status" : "fail",
+    "data" : { "Error" : "Error: Invalid argument:\nmoneymaker help \nfor more info." }
```
Comment (kinbiko) [inline design/api-specification-design.md]: Let's try and be kinder and let the user know what actually went wrong, since we don't want a 'help' endpoint. E.g. "amount missing"

**Review by kinbiko — CHANGES_REQUESTED** [review]
Good first draft. I like that you based your understanding of REST interfaces on one resource. The reason is that there are so many opinions on the matter, and trying to make sense and evaluate all of these is just going to waste your time. I've tried to make sure the comments I raised were not too contentious, opinion-wise.

1. I tried to help you out with JSON formatting. Markdown, and JSON in markdown is not usually a problem.
1. I think your content looks fine.
1. Yeah, PRs in the future would be good.

**Comment by perrymant** [comment]
I *think* I've addressed the improvements you have identified in this PR.

File: design/api-specification-design.md [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+╟────────────┼──────────────────┼────────┼─────────┼─────────────╢
+║ 2018-01-01 │ CREDIT           │ £1.25  │ £1.25   │ Got paid    ║
+╚════════════╧══════════════════╧════════╧═════════╧═════════════╝
+```
+
+- 500:  "Something went wrong. Please try again later.":
+
+```json
+{
+    "status" : "error",
```
Comment (kinbiko) [inline design/api-specification-design.md]: Still don't think this property adds any value.  Let's get rid of it throughout.

File: design/api-specification-design.md [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+
+```json
+{
+    "data" : {
+        "report" : "---report statement-(see below)---"
+     }
+}
+```
+
+An example `report statement` for status code 200 :
```
Comment (kinbiko) [inline design/api-specification-design.md]: So let's think about the purpose of this API, and APIs in general.

Right now we're saying that we should return this nicely formatted string of transactions. This makes quite a big assumption:

> The server decides how the client should present the data.

With this approach, the client has two options:

1. Display the data exactly in the format that the API gave it. Line endings and font spacing be damned.
1. Try and parse the data out of this custom string format into more manageable data structures.

This defeats the purpose of the problem that APIs try and solve. There's no need for the API to know how the data is presented. It only needs to care about the data itself, and organise it in a logically way. Like JSON!

See if you can come up with an alternative representation of this list of transactions in JSON.

Hint: We're somewhere between 80% and 100% of the way there with the work we did around persisting the transactions to a file.

File: design/api-specification-design.md [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+{
+    "status" : "error",
+    "message" : "Something went wrong. Please try again later."
+}
+```
+
+
+
+## 2. transaction
+A `transaction` request sends it to update their account.
```
Comment (kinbiko) [inline design/api-specification-design.md]: I don't understand this sentence...?

Is this closer to what you mean?
```suggestion
A `transaction` request is sent to append a new transaction to an account.
```

File: design/api-specification-design.md [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+| Method | URL             |
+|--------|-----------------|
+| POST   | api/transaction |
+
+#### Response
+
+- 200: A `transaction statement` will be sent:
+```json
+{
+  "transactions": [...]
```
Comment (kinbiko) [inline design/api-specification-design.md]: but what does this look like? Is it the same as `report`?

File: design/api-specification-design.md [hunk 16]
```diff
@@ -0,0 +1,83 @@
...
+}
+```
+
+- 400:  ">Error: Invalid argument: `moneymaker help` >for more info.":
```
Comment (kinbiko) [inline design/api-specification-design.md]: `moneymaker help` doesn't make sense in an API. You can just delete the sentence, or explain how this could happen in plain English.

File: design/api-specification-design.md [hunk 17]
```diff
@@ -0,0 +1,83 @@
...
+    "data" : { "Error" : "Error: amount missing" }
+}
+```
+
+- 500:  "Something went wrong. Please try again later.":
+
+```json
+{
+    "status" : "error",
+    "message" : "Unable to communicate with database"
```
Comment (kinbiko) [inline design/api-specification-design.md]: We shouldn't share this of information with the client for security reasons.

**Review by kinbiko — CHANGES_REQUESTED** [review]
Much better 👍

**Comment by perrymant** [comment]
1. `report` and `transaction` JSON updated
1. Makes more sense now about thinking _what_ a client should receive and have to deal with.

File: design/api-specification-design.md [hunk 18]
```diff
@@ -0,0 +1,86 @@
...
+|--------|-------------|
+| GET    | api/report  |
+
+#### Response
+- 200: a `report statement` will be sent:
+
+```json
+{
+    "data" : {
+        "report" : "---report statement-(see below)---"
```
Comment (kinbiko) [inline design/api-specification-design.md]: No need to delegate. Just include it here.

File: design/api-specification-design.md [hunk 19]
```diff
@@ -0,0 +1,86 @@
...
+     }
+}
+```
+
+An example `report statement` for status code 200 :
+```
+{
+"report": [
+  {
+    "date": ...,
```
Comment (kinbiko) [inline design/api-specification-design.md]: We're missing the good bits!
What's the format of the date? ISO8601 is always a good choice.
How is the amount and balance given? In pence or as a decimal? Or another format entirely?

File: design/api-specification-design.md [hunk 20]
```diff
@@ -0,0 +1,86 @@
...
+|--------|-----------------|
+| POST   | api/transaction |
+
+#### Response
+
+- 200: A `transaction statement` will be sent:
+```
+{
+"transaction": {
+    "date": ...,
```
Comment (kinbiko) [inline design/api-specification-design.md]: Same comment as for the report. Be explicit about this format.

File: design/api-specification-design.md [hunk 21]
```diff
@@ -0,0 +1,86 @@
...
+{
+    "data" : { "Error" : "The request contains bad syntax or cannot be fulfilled." }
+}
+```
+
+- 500:  "The server failed to fulfill an apparently valid request.":
+
+```json
+{
+    "message" : "The server failed to fulfill an apparently valid request."
```
Comment (kinbiko) [inline design/api-specification-design.md]: Perhaps it would be easier non-localised English for now, and change the wording to avoid certain words: "fulfill" (simplified English) vs fulfil (traditional English).

File: design/api-specification-design.md [hunk 22]
```diff
@@ -0,0 +1,86 @@
...
+}
+```
+
+
```
Comment (kinbiko) [inline design/api-specification-design.md]: Excessive whitespace

**Review by kinbiko — CHANGES_REQUESTED** [review]
Much better! Almost there now.

**Comment by perrymant** [comment]
- s/fulfill/completed
- formatted the JSON values
- removed excessive whitespace

File: design/api-specification-design.md [hunk 23]
```diff
@@ -11,40 +11,29 @@ The `report` request obtains a report of their account.
...
-"report": [
-  {
-    "date": ...,
-    "amount": ...,
-    "balance": ...,
-    "description": ...,
-  },
-  ...
-]
+"data" : {
```
Comment (kinbiko) [inline design/api-specification-design.md]: I still don't think we need this outer 'data' field.

Let's have an actual example here. You can document each field individually outside of this payload if you want to. Think of this as something you can copy and paste as a valid payload.

Also, let's be consistent with indentation (e.g. 2 spaces):

```json
{
  "report": [
    {
      "date": "2018-11-28T17:34:45Z",
      "amount": 1234,
      "balance": 12345678,
      "description": "blah"
    }
  ]
}
```

Applies below as well.

**Review by kinbiko — APPROVED** [review]
One more comment, then I think we're done.
